models/article.go

```diff
@@ -0,0 +1,209 @@
+	CreateArticle(*Article) error
+	DeleteArticle(*Article) error
+	GetAllArticles() *gorm.DB
+	GetAllArticlesAuthoredBy(string) ([]Article, error)
+	GetAllArticlesFavoritedBy(string) ([]Article, error)
+	GetAllArticlesWithTag(string) ([]Article, error)
+	GetArticle(string) (*Article, error)
+	FavoriteArticle(int, int) error
+	UnfavoriteArticle(int, int) error
+	FindUserByUsername(string) (*User, error)
+	IsFavorited(int, int) bool
+	IsFollowing(int, int) bool
+	SaveArticle(*Article) error
+}
+
+// Article article model
+type Article struct {
+	ID             int
+	Slug           string
+	Title          string
+	Description    string
+	Body           string
+	User           User
+	UserID         int
+	Tags           []Tag `gorm:"many2many:taggings;"`
+	Favorites      []Favorite
+	FavoritesCount int
+	CreatedAt      time.Time
+	UpdatedAt      time.Time
+}
+
+type ValidationMessages map[string]interface{}
+
+// NewArticle returns a new Article instance.
+func NewArticle(title string, description string, body string, user *User) *Article {
+	return &Article{
+		Title:       title,
+		Description: description,
+		Body:        body,
+		User:        *user,
+	}
+}
+
+// IsValid check if the article has a valid title, description and body
+func (a *Article) IsValid() (bool, map[string]interface{}) {
+	var errs = ValidationMessages{}
+	var valid = true
+
+	if a.Title == "" {
+		errs["title"] = []string{"title field can't be blank"}
+		valid = false
+	}
+
+	if a.Description == "" {
+		errs["description"] = []string{"description field can't be blank"}
+		valid = false
+	}
+
+	if a.Body == "" {
+		errs["body"] = []string{"body field can't be blank"}
+		valid = false
+	}
+
+	return valid, errs
+}
+
+func (a *Article) IsOwnedBy(username string) bool {
+	return a.User.Username == username
+}
+
+// CreateArticle persist a new article
+func (db *DB) CreateArticle(article *Article) (err error) {
+	err = db.Create(&article).Error
+	return
+}
+
+// DeleteArticle persist a new article
+func (db *DB) DeleteArticle(article *Article) (err error) {
+	err = db.Delete(&article).Error
+	return
+}
+
+func (db *DB) SaveArticle(article *Article) (err error) {
+	err = db.Save(&article).Error
+	return
+}
+
+// GetArticle retrieve an article by it slug
+func (db *DB) GetArticle(slug string) (*Article, error) {
+	var article Article
+	err := db.DB.Scopes(defaultScope).First(&article, "slug = ?", slug).Error
+	return &article, err
+}
+
+// GetAllArticles returns all articles.
+func (db *DB) GetAllArticles() *gorm.DB {
+	return db.DB.Scopes(defaultScope)
+}
+
+func (db *DB) GetAllArticlesWithTag(tagName string) (articles []Article, err error) {
+	//tag := Tag{Name: tagName}
+	//db.FindTag(&tag)
+	err = db. //DB.Model(&tag).
+			Scopes(defaultScope).
+			Joins("JOIN taggings ON taggings.article_id = articles.id").
+			Joins("JOIN tags ON tags.id = taggings.tag_id").
+			Where("tags.name = ?", tagName).
+			Find(&articles).Error
+	//Related(&articles, "Articles").Error
```

Comment:
I will remove It, it was my first attempt trying to understand gorm and use much of the DSL but it didn't feel natural, for example to get articles by tag I needed to fetch the tag then get its relationship to articles (for me its to many queries to do fo a simple task of fetching articles by tags so  I came up to build the query mixing the DSL and SQL and its more readable, WYSIWYG